tests/test_mal_simulator.py

```diff
@@ -956,6 +957,140 @@ def test_simulator_default_settings_defender_observation():
             assert not node.is_compromised() and not node.is_enabled_defense()
 
 
+def test_default_settings_defender_observation_false_negatives():
+    """Test default MalSimulator with false negative rates"""
+
+    sim, _ = create_simulator_from_scenario(
+        'tests/testdata/scenarios/traininglang_fp_fn_scenario.yml'
+    )
+    sim.reset()
+
+    attacker = sim.attack_graph.attackers[0]
+    attacker_agent_id = next(iter(sim.get_attacker_agents()))
+    defender_agent_id = next(iter(sim.get_defender_agents()))
+
+   # Get an uncompromised step
+    user_3_compromise = sim.attack_graph.get_node_by_full_name(
+        'User:3:compromise')
+    assert user_3_compromise.extras.get('false_negative_rate') == 1.0
+    assert attacker not in user_3_compromise.compromised_by
+
+    # Let the attacker compromise User:3:compromise
+    actions = {
+        attacker_agent_id: (1, sim._id_to_index[user_3_compromise.id]),
+        defender_agent_id: (0, None)
+    }
+
+    obs, _, _, _, _ = sim.step(actions)
+    defender_obs_state = obs[defender_agent_id]['observed_state']
+
+    # Verify that all states in obs match the state of the attack graph
+    node_index = sim._id_to_index[user_3_compromise.id]
+    # user_3_compromise is set to disabled even though compromised
+    # since it is false negative 100% of times
+    assert not defender_obs_state[node_index]
+
+
+def test_default_settings_defender_observation_false_negatives_seed():
+    """Test default MalSimulator with false negative rates"""
+
+    def run_simulation_with_seed(
+            seed,
+            expected_attacker_obs,
+            expected_defender_obs
+        ):
+        sim.reset(seed=seed)
+        attacker_agent_id = next(iter(sim.get_attacker_agents()))
+        defender_agent_id = next(iter(sim.get_defender_agents()))
+
+        # Get an uncompromised step
+        user_3_compromise = sim.attack_graph.get_node_by_full_name(
```

Comment:
This is good to check, but it would be even better if an actual attacker agent is used, or action_space.sample

Reply:
Do you mean to test the attacker agent seed as well? I think we should have a test for that separately.

Reply:
Perhaps both, even? One for the attacker and one for the integration of both the simulator and attacker.

Reply:
I added tests for the attacker agents now

Reply:
Added test for the combination of giving seed to agent AND simulator AND having fp/fn rates.